[Cleanup] Reduce build size to allow minimal_OTA builds for 1M

TD-er opened the pull request:
It links large blobs of data to online services like the Github CDN servers.
The CSS and JS files stored there will be served with a TTL of 1 year, so they should remain in the browser cache even if you're not connected to the internet, but once were.

Also added some more optimizations to finally make the minimal OTA builds again small enough for 2-step OTA updates.

Fixes: #3434

TD-er committed:
[Cleanup] Exclude unused code to reduce build size

TD-er committed:
[Cleanup] Reduce build size by using CDN for CSS and JS

TD-er committed:
[Cleanup] Reduce build size by using iterators for globalMapPortStatus

It also makes execution slightly faster.

TD-er committed:
Fix build issue for P019 (missing brace)

TD-er commented:
To make it all fit again in 1M builds, a lot of hours have been spent (over 100 hours)
So I would really appreciate if people could test the test build to see what I may have broken here.

- [ESPEasy_ESP32_mega-20201130-217-PR_3433.zip](https://www.dropbox.com/s/ml63wj691iqzsk9/ESPEasy_ESP32_mega-20201130-217-PR_3433.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20201130-217-PR_3433.zip](https://www.dropbox.com/s/mrtkydrb3f80c1a/ESPEasy_ESP82xx_mega-20201130-217-PR_3433.zip?dl=0)

TD-er commented:
Hmm aparently the builds made on Windows are a few kB smaller compared to those made on Linux :(
So the test build ZIP don't have the minimal OTA builds, while the builds made on Windows are all small enough to be used for minimal OTA on 1M units.

For now first trying to get some sleep as it is already half past late...

tonhuisman commented on static/toasting,js:
```diff
@@ -0,0 +1,9 @@
+function toasting() {
```
This filename `toasting,js` seems to have a comma for the extension instead of a period?

TD-er replied:
OK, that one I could not test yet as it is not yet included in the code with a tag

ironataerial commented:
Are there indeed ESP32 modules with 1M in the market ?

tonhuisman commented:
> Are there indeed ESP32 modules with 1M in the market ?

The smallest ESP32's I've seen had 2MB, AFAIR, but no, I don't think there are 1MB units actually available to the general public. Maybe on special order to save a few cents? But that's not going to make anyone happy 😉

TD-er commented:
I have not seen ESP32's with less than 4MB

TD-er committed:
Rename toasting,js to toasting.js

TD-er commented:
Ah got it...
The reported build size is not the final size as there's still some wappers being added.
Exactly the reason why I filter based on bin size after the build in the build script.
being hit by the same issue before and still not learning :(

TD-er committed:
[Cleanup] Reduce build size of WebPages

Lots of recurring nested function calls can be simplified via a convenience function.

TD-er committed:
[Cleanup] Disable "New Rules" on Limit build size builds

TD-er commented:
I updated the test build (and the links above)


- [ESPEasy_ESP32_mega-20201130-217-PR_3433.zip](https://www.dropbox.com/s/ml63wj691iqzsk9/ESPEasy_ESP32_mega-20201130-217-PR_3433.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20201130-217-PR_3433.zip](https://www.dropbox.com/s/mrtkydrb3f80c1a/ESPEasy_ESP82xx_mega-20201130-217-PR_3433.zip?dl=0)

Ladies and gentlemen... we've got'em!
There are minimal OTA builds for 1M included in those zip files.

ironataerial commented:
> I updated the test build (and the links above)
> 
> * [ESPEasy_ESP32_mega-20201130-217-PR_3433.zip](https://www.dropbox.com/s/ml63wj691iqzsk9/ESPEasy_ESP32_mega-20201130-217-PR_3433.zip?dl=0)
> * [ESPEasy_ESP82xx_mega-20201130-217-PR_3433.zip](https://www.dropbox.com/s/mrtkydrb3f80c1a/ESPEasy_ESP82xx_mega-20201130-217-PR_3433.zip?dl=0)
> 
> Ladies and gentlemen... we've got'em!
> There are minimal OTA builds for 1M included in those zip files.

Congrats man !! Unfortunately I do not have any 1M units to test but this took great effort !

TD-er committed:
Merge branch 'mega' into build/external_CSS_JS

TD-er committed:
Merge branch 'mega' into build/external_CSS_JS

TD-er committed:
[PIO] Add "display" and "energy" build collections

Both are based on the "normal" set, plus either all energy or all display related plugins included.

tonhuisman commented:
I've just tested these minimal builds a bit, but they have too much plugins missing for my use, so I think I'll probably have to improve my soldering skills (and equipment) so I can upgrade my Sonoff S20's and ESP01's to 4MB flash 😜, or use more Shelly's as most of them have 2MB Flash and the ESPEasy hardware builds allow OTA updates. But they are harder to add/fit extra sensors in the housing.
It is great to have these builds working again, but they are obviously for specific use-cases.

And now we also get these Display and Energy builds, that is useful 👍

tonhuisman commented:
Doesn't Core 2.7.x allow for compressed OTA? But there are no minimal builds that are available in compressed form. Shouldn't that allow for larger .bin files to be uploaded, or is that me, misunderstanding how that compression is supposed to work?

TD-er committed:
Merge branch 'mega' into build/external_CSS_JS

TD-er commented:
Compression only is useful during the flashing.
The active sketch must be stored uncompressed.

Thus these are both limited to the same max storage size:
- uncompressed sketch + compressed 2-step OTA
- uncompressed 2-step OTA + compressed new sketch

Initial tests have shown that when using core 2.7.x + SPIFFS support will result in a compressed file which is larger than the current 2-step OTA build.
So we can only use the compression when the 2-step OTA build does get its WiFi info from somewhere not stored in a file on SPIFFS.
For example when we use the EEPROM library (which we don't right now), we can store the current WiFi settings before starting the 2-step OTA update and this one should then also be able to read these WiFi settings.

TL;DR
It may help us, but needs some work to make it happen and the benefits are not that much.

TD-er commented:
Oh by the way @tonhuisman @ironataerial 
It would be nice if you could still test builds, not only the minimal OTA ones, but also the "test" builds, as those also use the same tricks to reduce binary size.

ironataerial commented:
I just updated one of the test units to ESP_Easy_mega_20201221_test_beta_ESP8266_4M1M from ESP_Easy_mega_20201130_normal_ESP8266_4M1M. 
Nothing funny so far. Will update if any issues arise.

tonhuisman commented:
I just updated 2 production units, both Wemos D1 mini clones, with this build, one has `test` (using APDS9660, OLedFramed and 2x Dummy Device plugins) and the other `normal` (only 4x Switch and 2x Dummy Device plugins, but a lot of rules).
No surprises so far (and that's good 👍).

TD-er committed:
[EmonCMS]  Allow sending 4 values per task (#3438)

Fixes: #3438

Also stores the values in the event queue element to make sure not only the last value of a task is being sent if sending values got delayed somehow.

TD-er committed:
[EmonCMS] Fix build issue

TD-er committed:
[Controller Queue] Optimize C004, C009 and C017

All were not storing the values to be queued so the values were just fetched from the task when the message was processed from the queue.
This means values could get lost if the queue was not processed for some time (or a task sends data very frequently)

TD-er commented:
New test build:

- [ESPEasy_ESP82xx_mega-20201130-276-PR_3433.zip](https://www.dropbox.com/s/gukakj2h8cxy41f/ESPEasy_ESP82xx_mega-20201130-276-PR_3433.zip?dl=0)
- [ESPEasy_ESP32_mega-20201130-276-PR_3433.zip](https://www.dropbox.com/s/b3g9hr10s27srl6/ESPEasy_ESP32_mega-20201130-276-PR_3433.zip?dl=0)

Main Changes:

Fixing queue handling for:  (values were not sent to the queue, only a marker with the taskIndex, thus only the last values when message was sent were used)

- C004: ThingSpeak
- C007: EmonCMS
- C009: FHEM HTTP
- C017: ZABBIX

Dallas:
When the (new) optional 2nd GPIO pin was not set, the Dallas sensor was not read. The sensors were found via a scan though.
This change is now being tested on my breadboard for over 24h already and is working just fine. Even with 12 of the cheapest Dallas sensors I could find online (based on the price, they must be couterfeit)

TD-er committed:
[PWM] Fix fade down (#3434)

Fixes: #3434

TD-er commented:
Now with the PWM also fixed (see #3434 for a test build), I plan on merging this PR tomorrow evening (well not "tomorrow", just the 26th)

So please let me know if there are other issues with the test builds presented here or in issue #3434 
Otherwise I will make a new nightly build tomorrow evening.
